src/Compilers/CSharp/Portable/FlowAnalysis/CSharpDataFlowAnalysis.cs

```diff
@@ -107,39 +107,33 @@ public override ImmutableArray<ISymbol> DataFlowsIn
         /// entered.
         /// </summary>
         public override ImmutableArray<ISymbol> DefinitelyAssignedOnEntry
-        {
-            get
-            {
-                if (_definitelyAssignedOnEntry.IsDefault)
-                {
-                    var entryResult = ImmutableArray<ISymbol>.Empty;
-                    var exitResult = ImmutableArray<ISymbol>.Empty;
-                    if (Succeeded)
-                    {
-                        var (entry, exit) = DefinitelyAssignedWalker.Analyze(_context.Compilation, _context.Member, _context.BoundNode, _context.FirstInRegion, _context.LastInRegion);
-                        entryResult = Normalize(entry);
-                        exitResult = Normalize(exit);
-                    }
-
-                    ImmutableInterlocked.InterlockedInitialize(ref _definitelyAssignedOnEntry, entryResult);
-                    ImmutableInterlocked.InterlockedInitialize(ref _definitelyAssignedOnExit, exitResult);
-                }
-
-                return _definitelyAssignedOnEntry;
-            }
-        }
+            => ComputeDefinitelyAssignedValues().onEntry;
 
         /// <summary>
         /// The set of local variables which are definitely assigned a value when a region is
         /// exited.
         /// </summary>
         public override ImmutableArray<ISymbol> DefinitelyAssignedOnExit
+            => ComputeDefinitelyAssignedValues().onExit;
+
+        private (ImmutableArray<ISymbol> onEntry, ImmutableArray<ISymbol> onExit) ComputeDefinitelyAssignedValues()
         {
-            get
+            if (_definitelyAssignedOnEntry.IsDefault)
             {
-                _ = DefinitelyAssignedOnEntry;
-                return _definitelyAssignedOnExit;
+                var entryResult = ImmutableArray<ISymbol>.Empty;
+                var exitResult = ImmutableArray<ISymbol>.Empty;
+                if (Succeeded)
+                {
+                    var (entry, exit) = DefinitelyAssignedWalker.Analyze(_context.Compilation, _context.Member, _context.BoundNode, _context.FirstInRegion, _context.LastInRegion);
+                    entryResult = Normalize(entry);
+                    exitResult = Normalize(exit);
+                }
+
+                ImmutableInterlocked.InterlockedInitialize(ref _definitelyAssignedOnEntry, entryResult);
```

Comment:
This looks like a race condition.

Thread 1 calls in, sees things are default, starts calculating, gets all the way to this line
Thread 2 class in, sees that `_definitelyAssignedOnEntry` is not default, returns `_definitelyAssignedOnExit`
Thread 1 sets `_definitelyAssignedOnExit`

Reply:
you are correct!  i am dumb.

Reply:
Ok.  i believe i have addressed this.  I'm not 100% certain, as we don't generally use this pattern on the IDE side.  But i feel like it shoudl be fine now as the new code says "only consider the cached results ok if both cached results are non-default."  LMK if i'm missing something subtle here.